Add asset and eth rate historical data

ddallaire opened the pull request:
- Add ethRate and assetRate historical values
  - Values are hooked to the existing rates updates instead of using a block handler
  - A reference is kept to the rate object to grab the currency and decimal places
  - Values are kept hourly to keep it lean
- Extract exchange rate objects to new file

jeffywu commented on src/notional.ts:
```diff
@@ -226,6 +219,8 @@ export function handleUpdateETHRate(event: UpdateETHRate): void {
   log.debug('Updated ethExchangeRate variables for entity {}', [ethExchangeRate.id]);
   ethExchangeRate.save();
+
+  updateEthExchangeRateHistoricalData(ethExchangeRate, event);
```
@ddallaire Putting this function call here won't do what you think it does, this `handleUpdateETHRate` will only be called rarely when the ETHRate configuration is updated by governance.

To get hourly updated exchange rate data you will need to register your function inside the subgraph template as a block handler:
https://thegraph.com/docs/en/developer/create-subgraph-hosted/#block-handlers

jeffywu commented on src/timeseriesUpdate.ts:
```diff
@@ -0,0 +1,60 @@
+        .concat(':')
+        .concat(uniqueHourIndex.toString());
+}
+
```
What I would do is put a function in here called `handleBlockUpdates` and call your `updateEthExchangeRate...` and `updateAssetExchangeRate...` methods every time `blockNumber % 276 == 0` or something to that effect. There are about 13 seconds between each block so you can assume that 276 blocks is about an hour.

jeffywu commented on src/timeseriesUpdate.ts:
```diff
@@ -0,0 +1,60 @@
+    let assetRate = rateResult.value3;
+
+    assetExchangeRateHistoricalData.timestamp = roundedTimestamp;
+    assetExchangeRateHistoricalData.value = assetRate.rate.div(BigInt.fromI32(10).pow(10));
```
I don't think you should divide by 10^10 here, let's leave the data as is and the client can do whatever division they want.

jeffywu commented on src/timeseriesUpdate.ts:
```diff
@@ -0,0 +1,60 @@
+    let roundedTimestamp = (event.block.timestamp.toI32() / 3600) * 3600;
+
+    let notional = Notional.bind(event.address);
+    let rateResult = notional.getCurrencyAndRates(event.params.currencyId);
```
As a consequence of calling this method via a blockHandler, you won't have access to the `event` object anymore. What I would recommend for getting the Notional address and the currencyId is this:

1. for the Notional address you can get it via `dataSource.address()` https://thegraph.com/docs/en/developer/assemblyscript-api/#data-source-metadata
2. for the currency id, you can call Notional.getMaxCurrencyId() and loop through currency id values where `1 <= currencyId <= maxCurrencyId` and get the data for each currencyId.

We should also store these two values:
notional.nTokenPresentValueAssetDenominated(currencyId)
notional.nTokenPresentValueUnderlyingDenominated(currencyId)

These methods may revert if there is no nToken present. In this case you do not need to store any value. You can wrap these calls in a try catch like this:
```
let pvAsset = notional.try_nTokenPresentValueAssetDenominated(currencyId)
let pvUnderlying = notional.try_nTokenPresentValueUnderlyingDenominated(currencyId)
if (pvAsset.reverted || pvUnderlying.reverted) return;
```

https://github.com/notional-finance/contracts-v2/blob/master/contracts/external/actions/nTokenAction.sol#L217-L233

ddallaire commented:
@jeffywu I updated the code to use a block handler. To make sure that we get at least 1 value per hour, I did not go with `blockNumber % 276` as the average has been [slightly more than 13 second per block](https://ycharts.com/indicators/ethereum_average_block_time) , I halved it, but I'm open to change that number.

I also added the historical values for `pvAsset` and `pvUnderlying`, let me know if that looks good.
